architecture/bela.cpp

```diff
@@ -625,48 +849,108 @@ class TrillCraftWidget : public TrillWidget
             mode = "PIN";
+            prescaler = 2;
+            threshold = 0.10;
+            start_note=0;
         }
         
-        virtual ~TrillCraftWidget()
-        {}
+        virtual ~TrillCraftWidget() 
+        {
+
+            for (auto t : Keyboard) {
+                delete t;
+            }
+
+        }
         
-        virtual void setParameters(const char* parameters)
+        virtual void setSensor(Trill* nSensor)
         {
-            // take the parameters
-            if (parameters) {
+            TrillWidget::setSensor(nSensor);   
+            if(sensor)
+            {
+                sensor->setPrescaler(prescaler);
+                sensor->setNoiseThreshold(threshold);
+            }
+        }
+        
+        virtual bool setParameters(const char* parameters)
+        {
+          
+           /*vector<string> names;
+            vector<double> values;
+            if(parseMenuList(parameters,names,values)) {
+                for(int i=0 ; i<names.size() ; i++) {
+                   setParameter(names[i],values[i]); 
+                }
+                
+            } else if (parameters) {*/
+            if(parameters && !TrillWidget::setParameters(parameters)) {
                 if (parseWord(parameters, "PIN")) {
-                    mode = "PIN";
+                    mode = "PIN"; 
                 } else if (parseWord(parameters, "UP")) {
                     mode = "UP";
                 } else if (parseWord(parameters, "DOWN")) {
                     mode = "DOWN";
                 }
+                double tmpval = 0.;
+                if (parameters && parseDouble(parameters, tmpval)) {
+                    lopin = (int)tmpval;
+                } else {
+                    lopin = -1;
+                }
+                if (parameters && parseChar(parameters, '-') && parseDouble(parameters, tmpval)) {
+                    hipin = (int)tmpval;
+                } else {
+                    hipin = -1;
+                }
             }
-            double tmpval = 0.;
-            if (parameters && parseDouble(parameters, tmpval)) {
-                lopin = (int)tmpval;
-            } else {
-                lopin = -1;
-            }
-            if (parameters && parseChar(parameters, '-') && parseDouble(parameters, tmpval)) {
-                hipin = (int)tmpval;
-            } else {
-                hipin = -1;
+            return true;
+        }
+        
+        virtual void setParameter(const string name, const double value)
+        {
+            TrillWidget::setParameter(name,value);
+            if(name == "prescaler") prescaler = (int) value;
+            else if(name == "threshold") threshold = (int) value;
+#ifdef NVOICES
+            else if(name == "start_pin") lopin = (int) value;
+            else if(name == "end_pin") hipin = (int) value;
+#endif  
+        }
+
+        virtual void setParameter(const string name, const string value)
+        {
+#ifdef NVOICES     
```

Comment:
Is conditional compilation really needs here ?

Reply:
no